readium/navigator/src/main/java/org/readium/r2/navigator/tts/AndroidTtsEngine.kt

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2022 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.navigator.tts
+
+import android.content.Context
+import android.content.Intent
+import android.speech.tts.TextToSpeech
+import android.speech.tts.UtteranceProgressListener
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.onFailure
+import org.readium.r2.navigator.tts.TtsEngine.*
+import org.readium.r2.shared.ExperimentalReadiumApi
+import org.readium.r2.shared.extensions.tryOrLog
+import org.readium.r2.shared.util.Language
+import org.readium.r2.shared.util.MapWithDefaultCompanion
+import org.readium.r2.shared.util.Try
+import java.util.*
+import kotlin.Exception
+import kotlin.coroutines.resume
+import android.speech.tts.Voice as AndroidVoice
+
+/**
+ * Default [TtsEngine] implementation using Android's native text to speech engine.
+ */
+@ExperimentalReadiumApi
+class AndroidTtsEngine(
+    context: Context,
+    private val listener: Listener
+) : TtsEngine {
+
+    /**
+     * Android's TTS error code.
+     * See https://developer.android.com/reference/android/speech/tts/TextToSpeech#ERROR
+     */
+    enum class EngineError(val code: Int) {
+        /** Denotes a generic operation failure. */
+        Unknown(-1),
+        /** Denotes a failure caused by an invalid request. */
+        InvalidRequest(-8),
+        /** Denotes a failure caused by a network connectivity problems. */
+        Network(-6),
+        /** Denotes a failure caused by network timeout. */
+        NetworkTimeout(-7),
+        /** Denotes a failure caused by an unfinished download of the voice data. */
+        NotInstalledYet(-9),
+        /** Denotes a failure related to the output (audio device or a file). */
+        Output(-5),
+        /** Denotes a failure of a TTS service. */
+        Service(-4),
+        /** Denotes a failure of a TTS engine to synthesize the given input. */
+        Synthesis(-3);
+
+        companion object : MapWithDefaultCompanion<Int, EngineError>(values(), EngineError::code, Unknown)
+    }
+
+    class EngineException(code: Int) : Exception("Android TTS engine error: $code") {
+        val error: EngineError =
+            EngineError.getOrDefault(code)
+    }
+
+    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
+
+    /**
+     * Utterances to be synthesized, in order of [speak] calls.
+     */
+    private val tasks = Channel<UtteranceTask>(Channel.BUFFERED)
```

Comment:
Using a `Channel` ensures that the utterances are played in the same order as the `speak()` calls.